transport/roundtrip_test.go

```diff
@@ -0,0 +1,152 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport_test
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/yarpc/yarpc-go/encoding/raw"
+	"github.com/yarpc/yarpc-go/transport"
+	"github.com/yarpc/yarpc-go/transport/http"
+	tch "github.com/yarpc/yarpc-go/transport/tchannel"
+	"github.com/yarpc/yarpc-go/transport/transporttest"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/testutils"
+)
+
+// all tests in this file should use these names for callers and services.
+const (
+	testCaller  = "testService-client"
+	testService = "testService"
+)
+
+// roundTripTransport provides a function that sets up and tears down an
+// Inbound, and provides an Outbound which knows how to call that Inbound.
+type roundTripTransport interface {
+	// Set up an Inbound serving Handler h, and call f with an Outbound that
+	// knows how to talk to that Inbound.
+	WithHandler(h transport.Handler, f func(transport.Outbound))
+}
+
+// handlerFunc wraps a function into a transport.Handler
+type handlerFunc func(context.Context, *transport.Request, transport.ResponseWriter) error
+
+func (f handlerFunc) Handle(ctx context.Context, r *transport.Request, w transport.ResponseWriter) error {
+	return f(ctx, r, w)
+}
+
+// httpTransport implements a roundTripTransport for HTTP.
+type httpTransport struct{ t *testing.T }
+
+func (ht httpTransport) WithHandler(h transport.Handler, f func(transport.Outbound)) {
+	i := http.NewInbound("127.0.0.1:0")
+	require.NoError(ht.t, i.Start(h), "failed to start")
+	defer i.Stop()
+
+	addr := fmt.Sprintf("http://%v/", i.Addr().String())
+	o := http.NewOutbound(addr)
+	f(o)
+}
+
+// tchannelTransport implements a roundTripTransport for TChannel.
+type tchannelTransport struct{ t *testing.T }
+
+func (tt tchannelTransport) WithHandler(h transport.Handler, f func(transport.Outbound)) {
+	serverOpts := testutils.NewOpts().SetServiceName(testService)
+	clientOpts := testutils.NewOpts().SetServiceName(testCaller)
+	testutils.WithServer(tt.t, serverOpts, func(ch *tchannel.Channel, hostPort string) {
+		i := tch.NewInbound(ch)
+		require.NoError(tt.t, i.Start(h), "failed to start")
+		defer i.Stop()
+		// ^ the server is already listening so this will just set up the
+		// handler.
+
+		client := testutils.NewClient(tt.t, clientOpts)
+		o := tch.NewOutbound(client, tch.HostPort(hostPort))
+
+		f(o)
+	})
+}
+
+func TestSimpleRoundTrip(t *testing.T) {
+	tests := []roundTripTransport{
+		httpTransport{t},
+		tchannelTransport{t},
+	}
+
+	request := &transport.Request{
+		Caller:    testCaller,
+		Service:   testService,
+		Procedure: "hello",
+		Encoding:  raw.Encoding,
+		Headers:   transport.Headers{"token": "1234"},
+		TTL:       200 * time.Millisecond, // TODO use default
+		Body:      bytes.NewReader([]byte("world")),
```

Comment:
If we later upgrade this test to also test different encodings. Right now it is just intended to test the happy case of different transports regardless of encoding. It doesn't do anything higher level.